lib/Dialect/FIRRTL/Transforms/DropName.cpp

```diff
@@ -0,0 +1,34 @@
+//===- DropName.cpp - Drop Names  -----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DropName pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct DropNamePass : public DropNameBase<DropNamePass> {
+  void runOnOperation() override {
+    getOperation()->walk([](FNamableOp op) {
+      if (!op.hasDroppableName())
```

Comment:
Since we're not yet defining O1, maybe we could start with this being a little more restricted? Could it check for `!op.hasDroppableName() && isOpDead(op)` where `isOpDead` is checking use empty or the op is only the destination of a connect?

It seems like a decent improvement anyway, and we can remove the `isOpDead` restriction after the `sv.namehint` improvements are in place. I'm just hoping we can achieve the cleanup we want here without causing a big regression in name preservation for non-dead wires.

Reply:
> Could it check for !op.hasDroppableName() && isOpDead(op) where isOpDead is checking use empty or the op is only the destination of a connect?

It is not impossible but I think it doesn't make sense from several reasons. If we really want to drop names of dead interesting wires, then I think we should rather delete operations transitively (the condition `!op.hasDroppableName() && isOpDead(op)` cannot handle well chains of dead operations).

> I'm just hoping we can achieve the cleanup we want here without causing a big regression in name preservation for non-dead wires.

Yes, I agree with this. TBH, I still don't think it is a bad option to use dead named wire elimination for a while until sv.namehint improves :sweat_smile:. It can be implemented by porting this commit https://github.com/llvm/circt/commit/622c66a14af2062f9041d1dd5ab7f9f8bece441b  into another dedicated pass.

Reply:
> then I think we should rather delete operations transitively

I see, but could that be "delete names transitively", rather than operations?

> Yes, I agree with this. TBH, I still don't think it is a bad option to use dead named wire elimination for a while until sv.namehint improves 😅. It can be easily implemented by porting this commit https://github.com/llvm/circt/commit/622c66a14af2062f9041d1dd5ab7f9f8bece441b into another dedicated pass.

I guess there are options on how it is implemented, but my hope is we can push forward so that the next release can include this PR and other work to remove dead named wires without causing a big name preservation regression. I guess we can measure the number of _GEN wires quantify if this regresses.

Reply:
Ok, I've implemented some sort of DCE and changed to drop names of dead operations.